File: skills/agent-patterns/templates/typescript/prompt_chaining.ts
/**
 * Prompt Chaining Pattern Implementation
 * Sequential LLM calls with programmatic checkpoints
 */

import Anthropic from "@anthropic-ai/sdk";

interface ChainStep<TContext = any> {
  name: string;
  promptTemplate: (context: TContext) => string;
  validator?: (output: string) => boolean;
  processor?: (output: string) => any;
}

interface ChainHistory {
  step: string;
  prompt: string;
  output: string;
  context: any;
}

export class PromptChain<TContext extends Record<string, any> = Record<string, any>> {
  private steps: ChainStep<TContext>[] = [];
  private history: ChainHistory[] = [];

  constructor(
    private client: Anthropic,
    private model: string = "claude-3-5-sonnet-20241022"
  ) {}

  /**
   * Add a step to the chain
   */
  addStep(step: ChainStep<TContext>): this {
    this.steps.push(step);
    return this; // Allow chaining
  }

  /**
   * Execute the chain with initial context
   */
  async execute(initialContext: TContext): Promise<string> {
    const context = { ...initialContext };
    let currentOutput = "";

    for (const step of this.steps) {
      // Format prompt with current context
      const prompt = step.promptTemplate(context as TContext);

      // Call LLM
      const message = await this.client.messages.create({
        model: this.model,
        max_tokens: 4096,
        messages: [{ role: "user", content: prompt }],
      });

      const textContent = message.content[0];
      if (textContent.type !== "text") {
        throw new Error("Expected text response");
      }
      currentOutput = textContent.text;

      // Validate if validator provided
      if (step.validator && !step.validator(currentOutput)) {
        throw new Error(
          `Step '${step.name}' validation failed. Output: ${currentOutput.substring(0, 100)}`
        );
      }

      // Process if processor provided
      if (step.processor) {
        const processed = step.processor(currentOutput);
        (context as any)[step.name] = processed;
      } else {
        (context as any)[step.name] = currentOutput;
      }

      // Track history
      this.history.push({
        step: step.name,
        prompt,
        output: currentOutput,
        context: { ...context },
      });
    }

    return currentOutput;
  }

  /**
   * Get execution history
   */
  getHistory(): ChainHistory[] {
    return this.history;
  }
}

// Example usage
interface DocumentContext {
  topic: string;
  outline?: string;
  draft?: string;
}

async function exampleDocumentGeneration() {
  const client = new Anthropic({
    apiKey: process.env.ANTHROPIC_API_KEY,
  });

  const chain = new PromptChain<DocumentContext>(client);

  // Step 1: Generate outline
  chain.addStep({
    name: "outline",
    promptTemplate: (ctx) => `Create a detailed outline for an article about: ${ctx.topic}`,
    validator: (output) => output.includes("1.") && output.includes("2."),
  });

  // Step 2: Expand outline
  chain.addStep({
    name: "draft",
    promptTemplate: (ctx) => `
      Expand this outline into a full article:
      ${ctx.outline}

      Write in a professional tone with clear examples.
    `,
    validator: (output) => output.split(" ").length > 200,
  });

  // Step 3: Proofread
  chain.addStep({
    name: "final",
    promptTemplate: (ctx) => `
      Proofread and polish this article:
      ${ctx.draft}

      Fix any grammar, improve clarity, and ensure consistent tone.
    `,
  });

  const result = await chain.execute({
    topic: "Building Effective AI Agents",
  });

  console.log("Final Article:");
  console.log(result);

  console.log("\n\nExecution History:");
  chain.getHistory().forEach((entry) => {
    console.log(`\nStep: ${entry.step}`);
    console.log(`Output length: ${entry.output.length} chars`);
  });
}

// Example with TypeScript generics for type safety
interface TranslationContext {
  topic: string;
  language: string;
  content?: string;
  terms?: string[];
}

async function exampleTypedChain() {
  const client = new Anthropic({
    apiKey: process.env.ANTHROPIC_API_KEY,
  });

  const chain = new PromptChain<TranslationContext>(client);

  chain.addStep({
    name: "content",
    promptTemplate: (ctx) => `Write a technical explanation of: ${ctx.topic}`,
  });

  chain.addStep({
    name: "terms",
    promptTemplate: (ctx) => `
      Extract technical terms from this text and define them:
      ${ctx.content}
    `,
    processor: (output) => {
      // Extract terms from output
      return output.split("\n").filter((line) => line.includes(":"));
    },
  });

  chain.addStep({
    name: "translation",
    promptTemplate: (ctx) => `
      Translate this text to ${ctx.language}, preserving these key terms: ${ctx.terms?.join(", ")}

      Text:
      ${ctx.content}
    `,
  });

  const result = await chain.execute({
    topic: "Quantum Computing",
    language: "Spanish",
  });

  return result;
}

// Export for use in other modules
export { exampleDocumentGeneration, exampleTypedChain };

File: skills/agent-patterns/templates/python/prompt_chaining.py
"""
Prompt Chaining Pattern Implementation
Sequential LLM calls with programmatic checkpoints
"""

from typing import Any, Callable, Optional
from dataclasses import dataclass
import anthropic


@dataclass
class ChainStep:
    """Represents a single step in the prompt chain"""
    name: str
    prompt_template: str
    validator: Optional[Callable[[str], bool]] = None
    processor: Optional[Callable[[str], Any]] = None


class PromptChain:
    """
    Executes a sequence of LLM calls with validation and processing between steps.

    Example:
        chain = PromptChain(client=anthropic_client)
        chain.add_step(
            name="outline",
            prompt_template="Create an outline for: {topic}",
            validator=lambda x: len(x.split('\n')) >= 3
        )
        chain.add_step(
            name="expand",
            prompt_template="Expand this outline into a full document:\n{outline}"
        )
        result = chain.execute({"topic": "AI Safety"})
    """

    def __init__(
        self,
        client: anthropic.Anthropic,
        model: str = "claude-3-5-sonnet-20241022"
    ):
        self.client = client
        self.model = model
        self.steps: list[ChainStep] = []
        self.history: list[dict[str, Any]] = []

    def add_step(
        self,
        name: str,
        prompt_template: str,
        validator: Optional[Callable[[str], bool]] = None,
        processor: Optional[Callable[[str], Any]] = None
    ) -> "PromptChain":
        """Add a step to the chain"""
        self.steps.append(ChainStep(name, prompt_template, validator, processor))
        return self  # Allow chaining

    async def execute(self, initial_context: dict[str, Any]) -> str:
        """Execute the chain with initial context"""
        context = initial_context.copy()
        current_output = ""

        for step in self.steps:
            # Format prompt with current context
            prompt = step.prompt_template.format(**context)

            # Call LLM
            message = await self.client.messages.create(
                model=self.model,
                max_tokens=4096,
                messages=[{"role": "user", "content": prompt}]
            )

            current_output = message.content[0].text

            # Validate if validator provided
            if step.validator and not step.validator(current_output):
                raise ValueError(
                    f"Step '{step.name}' validation failed. Output: {current_output[:100]}"
                )

            # Process if processor provided
            if step.processor:
                processed = step.processor(current_output)
                context[step.name] = processed
            else:
                context[step.name] = current_output

            # Track history
            self.history.append({
                "step": step.name,
                "prompt": prompt,
                "output": current_output,
                "context": context.copy()
            })

        return current_output


# Example usage
async def example_document_generation():
    """Example: Multi-step document generation"""
    client = anthropic.Anthropic()

    chain = PromptChain(client)

    # Step 1: Generate outline
    chain.add_step(
        name="outline",
        prompt_template="Create a detailed outline for an article about: {topic}",
        validator=lambda x: "1." in x and "2." in x  # Ensure numbered outline
    )

    # Step 2: Expand outline
    chain.add_step(
        name="draft",
        prompt_template="""
        Expand this outline into a full article:
        {outline}

        Write in a professional tone with clear examples.
        """,
        validator=lambda x: len(x.split()) > 200  # Ensure substantial content
    )

    # Step 3: Proofread
    chain.add_step(
        name="final",
        prompt_template="""
        Proofread and polish this article:
        {draft}

        Fix any grammar, improve clarity, and ensure consistent tone.
        """
    )

    result = await chain.execute({"topic": "Building Effective AI Agents"})
    return result


# Example with custom processing
async def example_with_processing():
    """Example: Translation pipeline with processing"""
    client = anthropic.Anthropic()

    def extract_key_terms(text: str) -> list[str]:
        """Simple key term extraction"""
        # In real implementation, use NLP or LLM
        return [word for word in text.split() if len(word) > 10]

    chain = PromptChain(client)

    chain.add_step(
        name="content",
        prompt_template="Write a technical explanation of: {topic}"
    )

    chain.add_step(
        name="terms",
        prompt_template="""
        Extract technical terms from this text and define them:
        {content}
        """,
        processor=extract_key_terms
    )

    chain.add_step(
        name="translation",
        prompt_template="""
        Translate this text to {language}, preserving these key terms: {terms}

        Text:
        {content}
        """
    )

    result = await chain.execute({
        "topic": "Quantum Computing",
        "language": "Spanish"
    })

    return result


if __name__ == "__main__":
    import asyncio
    asyncio.run(example_document_generation())
